bundler/lib/bundler/compact_index_client/updater.rb

```diff
@@ -31,6 +31,11 @@ def update(local_path, remote_path, retrying = nil)
 
           # first try to fetch any new bytes on the existing file
           if retrying.nil? && local_path.file?
+            # Ensure that we can read the source since we can't distinguish an
+            # access error between the source and target file in FileUtils.cp.
+            SharedHelpers.filesystem_access(local_path, :read) do
```

Comment:
In general, I find `SharedHelpers.filesystem_access` problematic because they don't check whether the `EACCES` error comes from the path they're supposed to protect. Nested calls can attribute the exception to the wrong path.